format-all.el

```diff
@@ -1015,21 +1018,36 @@ unofficial languages IDs are prefixed with \"_\"."
                   executable
                   (gethash formatter format-all--install-table)))))))
 
+(defun format-all--show-errors-buffer (error-output show-errors-p)
+  "Internal shorthand function to update and show error output.
+
+ERROR-OUTPUT come from the formatter.  SHOW-ERRORS-P determines
+whether or not to display the errors buffer."
+  (save-selected-window
+    (with-current-buffer (get-buffer-create "*format-all-errors*")
+      (erase-buffer)
+      (insert error-output)
+      (if show-errors-p
+          (display-buffer (current-buffer))
+        (let ((error-window (get-buffer-window (current-buffer))))
+          (when error-window (quit-window nil error-window)))))))
+
 (defun format-all--update-errors-buffer (status error-output)
   "Internal helper function to update *format-all-errors*.
 
 STATUS and ERROR-OUTPUT come from the formatter."
-  (let ((show-errors-p (and (not (= 0 (length error-output)))
-                            (or format-all-always-show-errors
-                                (eq status :error)))))
-    (save-selected-window
-      (with-current-buffer (get-buffer-create "*format-all-errors*")
-        (erase-buffer)
-        (insert error-output)
-        (if show-errors-p
-            (display-buffer (current-buffer))
-          (let ((error-window (get-buffer-window (current-buffer))))
-            (when error-window (quit-window nil error-window))))))))
+  (let* ((show-errors-choice (symbol-name format-all-show-errors))
+         (has-warnings-p (not (= 0 (length error-output))))
+         (has-errors-p (eq status :error))
+         (show-errors-p (and (not (string-equal show-errors-choice 'never))
+                             (or
+                              (and (string-equal show-errors-choice 'always)
+                                   (or has-errors-p has-warnings-p))
+                              (and (string-equal show-errors-choice 'warnings)
+                                   (or has-errors-p has-warnings-p))
+                              (and (string-equal show-errors-choice 'errors)
+                                   has-errors-p)))))
+    (format-all--show-errors-buffer error-output show-errors-p)))
 
```

Comment:
No need to use `symbol-name` - symbols can be tested simply by using `(eq 'foo 'foo)`. In fact, we can use a case statement:

```
(cl-case format-all-show-errors
  (always ...)
  (warnings ...)
  (errors ...)
  (never ...)
  (otherwise ...))
```

Reply:
thanks, it definitely clears the code a lot.

Reply:
Looks great!